Revision of Bootswatch themes

tarvos21 opened the pull request:
- [x] cerulean            
- [x] cosmo           
- [x] cyborg            
- [x] darkly            
- [x] flatly            
- [x] journal           
- [x] lumen           
- [x] paper           
- [x] readable            
- [x] sandstone           
- [x] simplex           
- [x] slate           
- [x] spacelab            
- [x] superhero           
- [x] united            
- [x] yeti

## Cerulean
- Global
  - **header and page font-family**: from default Lato to Helvetica Neue
- Elements
  - **button**: background and box-shadow effects modified
  - **step, breadcrumb**: background color
- Collections
  - **menu**: active item's text color, from black to white
  - **vertical and tabular menu**: keep text color consistent with other menu
## Cosmo
- Global
  - **global colors**: redefine variable names
- Elements
  - **button**: basic button's text visibility
  - **step**: background color
  - **label**: text color
- Collections
  - **breadcrumb**: background color
  - **form**: warning, error form input's border, background color modified
  - **menu**: text color(hover, active)
  - **message**: header, description's text color
  - **table**: positive, negative, warning table row and cell's background color
- Modules
  - **accordion**: make accordion menu appear right
## Cyborg
- Global
  - **code color**: improve visibility on black background 
  - **state colors**: hovered, pressed, selected, disabled, unselected
  - **placeholder color**: from white to gray 
- Elements
  - **code icon**: from black to gray
  - **divider**: color of text inside the divider 
  - **block and attached header**: background color
  - **corner label**: top and right border 
  - **list**: background set to transparent to keep consistent 
  - **list description**: text color 
  - **piled and stacked segment**: decorative segments color, from white to dark 
  - **step**: add a style, almost same as cosmo theme 
- Collections
  - **breadcrumb**: background set to transparent, and make text visible 
  - **form**: warning, error form input's border, background color modified
  - **vertical menu**: background set to transparent 
  - **table**: remove box-shadow of the empty header in definition table 
- Views
  - **feed**: color of meta, date, extra parts 
  - **statistic**: big number's color, from gray to white
- Modules
  - **checkbox**: toggle and slider checkbox's slider bar, from black to gray 
  - **rating icon**: color when inactive
## Darkly
- Global
  - **default and inverted color**: exchange, and do some modification 
  - **code color**: improve visibility on black background 
  - **state background colors**: hovered, pressed, selected, disabled, unselected
  - **border color**: gray 
  - **placeholder color**: from white to gray 
- Elements
  - **code icon**: gray, more visible
  - **divider**: color of text inside the divider 
  - **block and attached header**: background color
  - **input**: border and height 
  - **list**: background set to transparent to keep consistent 
  - **list description**: text color 
  - **piled and stacked segment**: decorative segments color, from white to dark 
  - **step**: add a style, almost same as cosmo theme 
- Collections
  - **breadcrumb**: background set to transparent, and make text visible 
  - **form**: warning, error form input's border, background color modified
  - **menu**:
    - sub menu's text color
    - vertical menu's background, from white to none
    - hovered color of dropdown menu's item
    - default and active background of secondary menu's item
    - active background of tabular menu's item
  - **message**: default background color 
  - **table background colors**:
    - default
    - header
    - footer
    - definition table
    - active row or cell
- Views
  - **feed**: color of meta data and label icon, increase visibility
  - **statistic**: value and label color, big contrast 
- Modules
  - **checkbox**: toggle and slider checkbox's slider bar, from transparent to gray 
  - **dropdown**: header and divider's colors 
  - **progress bar**: consistent height of 10px 
  - **rating icon**: visible color when inactive
## Flatly
- Global
  - **state background colors**: selected, muted
  - **placeholder color**: increase visibility
  - **solid definition**: avoid references 
- Elements
  - **input label**: make the text vertically center aligned when height increased
  - **corner label**: icon's color 
  - **piled and stacked segment**: decorative segments color, from white to light gray 
  - **step**: add a style
- Collections
  - **breadcrumb**: increase padding, but background not changed 
  - **form**: warning, error form input's border, background color modified
  - **menu**:
    - active item text color in rail vertical menu 
    - vertical menu's background
    - tabular menu's default text color 
  - **table background colors**: remove color reference in global setting, then work
- Modules
  - **progress bar**: consistent height of 10px, use variable to define
## Journal
- Global
  - **font family**: modify a few variables
    - google font must be open source
- Elements
  - **button**: delete duplicate text color variable
## Lumen
- Global
  - **global colors**: remove reference, use direct definition 
  - **google fonts**: use only available open source font 
- Elements
  - **button**: 
    - font-size: smaller than page font size
    - text-transform: all uppercase
    - border: border-bottom is emphasized
  - **input**: height, border, label middle aligned 
  - **segment, step**: add border style 
- Collections
  - **form**: remove inconsistent border 
  - **menu**: 
    - border-bottom: 4px
    - active item background color(darker)
  - **message**: add border style
## Paper
- Global
  - **global colors**: avoid variables references 
  - **headers**: use relative font size
  - **font family**: remove superfluous fonts 
- Elements
  - **button**:
    - remove useless lines
    - effects when button is clicked
    - border
    - uppercase text
  - **input**: bigger line height, visible border 
  - **segment header**: remove border 
  - **step**: border 
- Collections
  - **breadcrumb**: increase padding, add border 
  - **form**: 
    - input and select border and height
    - error field input text color
  - **menu**: border of icon
  - **message**: default text color 
- Views
  - **card**: add border
- Modules
  - **progress**: bar height
## Readable
- Global
  - **global colors**: remove variables references
  - **global font**: remove incorrect definition
- Elements
  - **button**: remove @fontName 
  - **input**: correct border definition 
  - **label**: background color and corner label's border 
- Collections
  - **menu**: minor revision of border
## Sandstone
- Global
  - **global colors**: remove variables references
  - **global font**: remove incorrect definition
- Elements
  - **basic button**: text color, uppercase 
  - **input**: correct border definition 
  - **label**: text and background color 
  - **segment**: remove border and text-transform settings 
  - **step**: add text-transform setting to title
- Collections
  - **breadcrumb**:
    - exchange default and active section text color
    - increase padding
  - **menu**: vertical menu's label background, active text color
  - **form**: error field input placeholder color, more visible
## Simplex
- Global
  - **global font**: remove duplicate definition
- Elements
  - **button**:
    - default hovered text color
    - basic button text color, set to black
  - **input**: make border visible 
  - **step**: add border and background style
## Slate
- Global
  - **default and inverted color**: exchange, and do some modification 
  - **code color**: improve visibility on black background 
  - **border color**: gray 
  - **placeholder color**: from white to gray 
- Elements
  - **button**:
    - basic button: background changed to gray
    - add gradient effects to button background
    - change hover background color
  - **block and attached header**: background color
  - **code icon**: gray, more visible
  - **input**: border, placeholder color 
  - **list**:
    - description text color, more visible
    - make text color of links white in inverted list(page footer)
    - background set to transparent to keep consistent 
  - **piled and stacked segment**: decorative segments color, from white to dark 
  - **step**: add a style 
- Collections
  - **breadcrumb**:
    - increase padding
    - default and active text color
  - **form**:
    - make text of select visible
    - add gradient effects to label and button in inputs
  - **menu**:
    - remove inconsistent border style
    - hovered color of dropdown menu's item
    - sub menu's text color
    - vertical menu's background, label background
    - pointing border color of pointing secondary menu
  - **message**: default background color 
  - **table background colors**:
    - default
    - header
    - footer
    - definition table border
    - active row or cell
- Views
  - **feed**: color of meta data
  - **statistic**: value color 
- Modules
  - **checkbox**: toggle and slider checkbox's slider bar, bullet color of checked radio 
  - **rating icon**: set to visible color when inactive
## Spacelab
- Global
  - **font and color**: minor revision of variables 
- Elements
  - **button**:
    - basic button text color
    - hovered background
    - background of button in input
  - **step**: add gradient background
## Superhero
- Global
  - **default and inverted color**: exchange, and do some modification 
  - **code color**: improve visibility on black background 
  - **border color**: gray 
  - **placeholder color**: more visible and consistent
- Elements
  - **basic button**: hover background color
  - **block and attached header**: consistent background color
  - **code icon**: gray, more visible
  - **input**: border width set to zero 
  - **label**: background set to @primaryColor 
  - **list**:
    - description text color, more visible
    - background set to transparent to keep consistent 
  - **piled and stacked segment**: decorative segments color, from white to gray 
  - **step**: add a style 
- Collections
  - **breadcrumb**: increase padding
  - **form**:
    - make text of select visible
    - input placeholder color modified 
  - **menu**:
    - set text and background color of dropdown menu
    - active hovered item's text color
    - vertical menu's background and label background color
    - tabular and secondary active item background color
  - **message**: default text and background color 
  - **table background colors**:
    - default
    - header
    - footer
    - striped
    - definition table border
    - active row or cell
- Views
  - **feed**: meta data and icon color
  - **statistic**: value and label color
- Modules
  - **checkbox**: toggle and slider checkbox's slider bar, bullet color of checked radio 
  - **rating icon**: set to visible color when inactive
## United
- Global
  - **font**: set to Ubuntu 
  - **muted text color**: set to light gray(for text menu item)
- Elements
  - **basic button**: default text color
  - **step**: add a style 
- Collections
  - **menu**:
    - secondary menu default text color
    - tabular menu deafult text color
    - hover color of active item
    - vertical menu background, primary color
    - label background in menu
## Yeti
- Global
  - **font**: simplify definition 
- Elements
  - **button**: unify padding 
  - **step**: add theme style, black
- Collections
  - **breadcrumb**: increase padding
  - **menu**:
    - secondary menu's default text color
    - accordion vertical menu's background color, transparent
    - label background in menu
    - hover text and background color of active item
    - text and background color of dropdown
    - vertical menu's default background
    - tabular menu's default text color
- Modules
  - **dropdown**: header text color
  - **popup**: remove previous background color

tarvos21 committed:
cerulean: active item text color, vertical and tabular menu text color

tarvos21 committed:
cerulean: button's background and box-shadow effects

tarvos21 committed:
cerulean: header, page font-family definition

xiaohanyu commented on src/themes/cerulean/globals/site.variables:
```diff
@@ -19,3 +19,5 @@
 @negativeTextColor: #b94a48;
 @negativeBorderColor: darken(spin(@negativeBackgroundColor, -10), 3%);
+
+@fontName: 'Helvetica Neue';
```
Default `site.variables`:  https://github.com/semantic-ui-forest/semantic-ui/blob/master/src/themes/default/globals/site.variables#L9-L21

```
@fontName          : 'Lato';
@fontSmoothing     : antialiased;

@headerFont        : @fontName, 'Helvetica Neue', Arial, Helvetica, sans-serif;
@pageFont          : @fontName, 'Helvetica Neue', Arial, Helvetica, sans-serif;

@googleFontName    : @fontName;
@importGoogleFonts : true;
@googleFontSizes   : '400,700,400italic,700italic';
@googleSubset      : 'latin';

@googleProtocol    : 'https://';
@googleFontRequest : '@{googleFontName}:@{googleFontSizes}&subset=@{googleSubset}';
```

I think if one theme does not use Google font, then we should set `@importGoogleFonts : true;` to `@importGoogleFonts : false;`.

And, besides set `@fontName: 'Helvetica Neue'`, it's better to modify `@headerFont` and `@pageFont` to remove 'Helvetica Neue' duplicate.

tarvos21 replied:
done.

xiaohanyu commented on src/themes/cerulean/elements/button.overrides:
```diff
@@ -1,23 +1,45 @@
-.ui.primary.button {
-  background-image: @primaryBackgroundImage;
-  border-bottom: 1px solid darken(@primaryColor, 10%);
+.btn-shadow(@color) {
```
`.btn-shadow` ?

xiaohanyu commented on src/themes/cerulean/collections/menu.overrides:
```diff
@@ -1,7 +1,79 @@
 
-.ui.tabular.menu .item {
-  color: @secondaryItemColor;
+.btn-shadow(@color) {
```
What's `.btn-shadow` do here?

tarvos21 replied:
It's like a function, including style definitions, and the `#gradient { .vertical-three-colors ...}` is like a class definition, and `.btn-shadow` uses a function in that class. This is a feature of LESS language, it's powerful, but also confusing when I first encounter.

tarvos21 replied:
https://github.com/thomaspark/bootswatch/blob/gh-pages/bower_components/bootstrap/less/mixins/vendor-prefixes.less
You can find many this kinds of "function"s here.

tarvos21 replied:
Because I do not understand this when I process the Cerulean theme, so I write the style directly into the `*.overrides` and `*.variables`, to make it work right. After 2-3 themes, I find this LESS common feature and understand what it does, but at last I forget to modify the Cerulean theme in this way. Sorry for this.

xiaohanyu replied:
So I guess it's a mixin in LESS/SASS language.

It's OK to use this. One thing I prefer, please use full word instead of abbreviations, just to make it more like semantic-ui's `ui button` instead of bootstrap's `btn-`. For `btn-shadow`, prefer `button-shadow`.

tarvos21 replied:
OK, I will do that with cute `sed`.

xiaohanyu reviewed (changes requested):
Please modify

tarvos21 committed:
cerulean: breadcrumb's background color

tarvos21 committed:
ceruelan: step's background-color

tarvos21 committed:
replace .btn-shadow with .button-shadow in all files

tarvos21 committed:
update cerulean.theme.config

tarvos21 committed:
cosmo: label's text color

tarvos21 committed:
cosmo: step's background

tarvos21 committed:
cosmo: menu's text color(hover, active)

tarvos21 committed:
cosmo: form input's border, background

tarvos21 committed:
cosmo: message's header and description text color

tarvos21 committed:
cosmo: positive, negative, warning table background

tarvos21 committed:
cosmo: accordion menu's colors

tarvos21 committed:
cosmo: global colors definition

tarvos21 committed:
update cosmo.theme.config

tarvos21 committed:
cosmo: basic button text color

tarvos21 committed:
update cyborg.theme.config

tarvos21 committed:
cyborg: text color inside divider

tarvos21 committed:
cyborg: code icon's color

tarvos21 committed:
cyborg: block and attached header's background color

tarvos21 committed:
cyborg: corner label's border

tarvos21 committed:
cyborg: list description's text color

tarvos21 committed:
cyborg: list background set to transparent

tarvos21 committed:
cyborg: piled and stacked segment's decorative segments' color

tarvos21 committed:
cyborg: step element, background, text color

tarvos21 committed:
cyborg: breadcrumb's theme

tarvos21 committed:
cyborg: box-shadow, background of elements in form

tarvos21 committed:
cyborg: vertical menu's background

tarvos21 committed:
cyborg: definition table's empty thead's box-shadow

tarvos21 committed:
cyborg: toggle and slider checkbox's slider bar color

tarvos21 committed:
cyborg: inactive rating icon's color

tarvos21 committed:
cyborg: feed and statistic, meta color, value color

tarvos21 committed:
cyborg: generated code's text color

tarvos21 committed:
cyborg: global colors setting

- input placeholder's color
- disabled text color
- hovered, pressed, selected and unselected color
- some revisions of variables definition

tarvos21 committed:
update darkly.theme.config

tarvos21 committed:
darkly: divider's color

tarvos21 committed:
darkly: make code icon visible on dark page

tarvos21 committed:
darkly: input's border, height, label's lineheight

tarvos21 committed:
darkly: block and attached header's background

tarvos21 committed:
darkly: remove label's superfluous variable definition

tarvos21 committed:
darkly: list's background and description color

tarvos21 committed:
darkly: step background color, active, default, disabled

tarvos21 committed:
darkly: form's input

- input border
- input text color
- placeholder color

tarvos21 committed:
darkly: breadcrumb's background and text color

tarvos21 committed:
darkly: various menu's background and text color

- sub menu's text color
- vertical menu's background, from white to none
- hovered color of dropdown menu's item
- default and active background of secondary menu's item
- active background of tabular menu's item

tarvos21 committed:
darkly: message's default background color

tarvos21 committed:
darkly: table rows' background

- default background color
- header
- footer
- definition table
- active row or cell

tarvos21 committed:
darkly: meta data and label icon color in feed

tarvos21 committed:
darkly: value and label color in statistic

tarvos21 committed:
darkly: piled and stacked segment's decorative segment background

tarvos21 committed:
darkly: slider and toggle checkbox's slider bar and bullet color

tarvos21 committed:
darkly: dropdown's header and divider color

tarvos21 committed:
darkly: redefine progress bar's height

tarvos21 committed:
darkly: inactive rating icon's color

tarvos21 committed:
darkly: global colors settings

- default and inverted colors
- background color of different states
- border color
- input and placeholder color
- generated code's text color, more visible
- modify suspicious variable definitions

tarvos21 committed:
update flatly.theme.config

tarvos21 committed:
flatly: remove useless text color definition

tarvos21 committed:
flatly: input label text vertically center aligned

- label's line height same as input
- vertical-align: middle

tarvos21 committed:
flatly: corner label's icon color

tarvos21 committed:
flatly: piled and stacked segment's decorative segment color

tarvos21 committed:
flatly: add background color to step

tarvos21 committed:
flatly: increase breadcrumb's padding

tarvos21 committed:
flatly: modify input border and background in form

tarvos21 committed:
flatly: vertical menu, tabular menu color and background

- vertical menu in rail, white
- active item text color, black
- tablular menu's text color, make it visible

tarvos21 committed:
flatly: redefine progress bar height

tarvos21 committed:
flatly: solid global colors settings

tarvos21 committed:
global font settings modified

Checked themes including: cerulean, cosmo, cyborg, darkly, flatly
- @importGoogleFonts: set to false if not needed
- put more priority on @pageFont and @headerFont

Same problems will be handled respectively in following themes

tarvos21 committed:
update journal.theme.config

tarvos21 committed:
journal: modify a few variables (including google font)

special things about google font request:

- mutiple fonts: 'Font-A|Font-B|Font-C'
- must be open source font

tarvos21 committed:
update lumen.theme.config

tarvos21 committed:
lumen: button's padding, font size, border

- padding: consistent among different buttons
- border: 0 1px 4px 1px
- font-size: base is 12px

tarvos21 committed:
lumen: input border, height, label

- make borer visible
- keep label center aligned vertically
- corner label input: icon display normally

tarvos21 committed:
lumen: add border-bottom to segment

tarvos21 committed:
lumen: add border-bottom to step

tarvos21 committed:
lumen: remove form's wrong border

tarvos21 committed:
lumen: menu's border, active item background

tarvos21 committed:
lumen: message border-bottom color

tarvos21 committed:
lumen: add border-bottom to card

tarvos21 committed:
lumen: global colors, remove references in definition

tarvos21 committed:
paper: refactor button's overrides a lot

- remove uselsss lines
- effects when button is clicked
- border
- uppercase

tarvos21 committed:
paper: input's line height, border

tarvos21 committed:
paper: step border

tarvos21 committed:
paper: let header in segment have border

tarvos21 committed:
paper: breadcrumb's border, padding

tarvos21 committed:
paper: form elements' border, color

- input and select border and height
- error field input text color

tarvos21 committed:
paper: menu element border

tarvos21 committed:
paper: message default text color

tarvos21 committed:
paper: add border to card

tarvos21 committed:
paper: bar height of progress

tarvos21 committed:
paper: global variables

- header's font size(relative)
- state colors
- font family

tarvos21 committed:
update readable.theme.config

tarvos21 committed:
readable: remove superfluous font variable of button

tarvos21 committed:
readable: remove improper border of input

tarvos21 committed:
readable: label background and corner label's border

tarvos21 committed:
readable: menu border

tarvos21 committed:
readable: global variables minor revision

tarvos21 committed:
update sandstone.theme.config

tarvos21 committed:
sandstone: basic button text color, text transform

tarvos21 committed:
sandstone: remove text-transform for header

tarvos21 committed:
sandstone: remove input border setting

tarvos21 committed:
sandstone: label's text and background color

tarvos21 committed:
sandstone: remove segment's border and text-transform setting

tarvos21 committed:
sandstone: set title in step to uppercase

tarvos21 committed:
sandstone: breadcrumb default and active text color, padding

tarvos21 committed:
sandstone: vertical menu's label background, active text color

tarvos21 committed:
sandstone: form error field input placeholder color

tarvos21 committed:
sandstone: redefine colors and fonts

tarvos21 committed:
update simplex.theme.config

tarvos21 committed:
simplex: basic button text color, button hovered text color

tarvos21 committed:
simplex: make input's border visible

tarvos21 committed:
simplex: add border and backbground style for step

tarvos21 committed:
simplex: minor modification of global variables

tarvos21 committed:
update slate.theme.config

tarvos21 committed:
slate: button hover background, basic button background

- Add ".ui" to ".button": bring background style into effect

tarvos21 committed:
slate: set code icon color particularly

tarvos21 committed:
slate: correct input's border and placeholder color

tarvos21 committed:
slate: make block and attached header background consistent

tarvos21 committed:
slate: list description, and inverted link text color

- remove improper and unsafe styles

tarvos21 committed:
slate: border and background of different segments

- background decorative segments color, of stacked and attached segment
- change border color to gray, more gentle

tarvos21 committed:
slate: add style to step element

tarvos21 committed:
slate: breadcrumb default and active text color, padding

tarvos21 committed:
slate: form's select color, label and button input background

- select's text color
- label and button input background, gradient effect

tarvos21 committed:
slate: menu's border, background, text color

- remove inconsistent border style
- set text and background color of dropdown menu
- active hovered item's text color, changing a little
- vertical menu's background and label background color
- pointing border color of pointing secondary menu

tarvos21 committed:
slate: background color of default message

tarvos21 committed:
slate: table background of header, footer, states, active row

tarvos21 committed:
slate: checkbox slider bar color, bullet color

- bullet color when checked

tarvos21 committed:
slate: rating default icon color

tarvos21 committed:
slate: feed meta data color

tarvos21 committed:
slate: statistic value color

tarvos21 committed:
slate: global colors setting

- default and inverted colors swapped
- remove references in variable definitions
- make code more visible

tarvos21 committed:
update spacelab.theme.config

tarvos21 committed:
spacelab: button hover text color, basic button text color

- input button background

tarvos21 committed:
spacelab: add background style to step

tarvos21 committed:
spacelab: minor revison of global font and color variables

tarvos21 committed:
update superhero.theme.config

tarvos21 committed:
superhero: basic button hover background color

tarvos21 committed:
superhero: make block and attached header background consistent

tarvos21 committed:
superhero: set code icon color particularly

tarvos21 committed:
superhero: input boder width set to zero

tarvos21 committed:
superhero: list description text color and background

tarvos21 committed:
superhero: label's background set to @primaryColor

tarvos21 committed:
superhero: background color of stacked and piled segments' decorative segments

tarvos21 committed:
superhero: add style to step element

tarvos21 committed:
superhero: increase breadcrumb's padding

tarvos21 committed:
superhero: form input border color, select text color

tarvos21 committed:
superhero: various menus' background, text color

- set text and background color of dropdown menu
- active hovered item's text color
- vertical menu's background and label background color
- tabular and secondary active item background color

tarvos21 committed:
superhero: make message text color and background color consistent

tarvos21 committed:
superhero: table background of header, footer, states, active row

tarvos21 committed:
superhero: feed icon and meta data color

tarvos21 committed:
superhero: statistic value and label color

tarvos21 committed:
superhero: checkbox slider and toggle bar, bullet color

tarvos21 committed:
superhero: rating icon default color

tarvos21 committed:
superhero: global colors setting

- default and inverted colors swapped
- make code more visible
- remove superfluous font definition
- remove references in variable definitions
- border, placeholder text color

tarvos21 committed:
update united.theme.config

tarvos21 committed:
united: modify basic button text color

tarvos21 committed:
united: add a style to step element

tarvos21 committed:
united: various menus' text and background color

- secondary menu default text color
- tabular menu deafult text color
- hover color of active item
- vertical menu background, primary color
- label background in menu

tarvos21 committed:
united: minor modification of global font and color

tarvos21 committed:
slate: popup background modified

tarvos21 committed:
yeti: make button's padding consistent

tarvos21 committed:
yeti: remove label's background setting

tarvos21 committed:
yeti: add theme style to step element

tarvos21 committed:
yeti: increase breadcrumb padding

tarvos21 committed:
yeti: various menu's text and background color

- secondary menu's default text color
- accordion vertical menu's background color, transparent
- label background in menu
- hover text and background color of active item
- text and background color of dropdown
- vertical menu's default background
- tabular menu's default text color

tarvos21 committed:
yeti: header text color in dropdown

tarvos21 committed:
yeti: remove popup background color

tarvos21 committed:
yeti: simplify global font setting

tarvos21 commented:
A little reflection of the second round of theme porting work

Typically, I check the themes one by one to find problems for each theme, and each theme may have different numbers of problems, something like this:

|  |  |  |  |  |
--- | --- | --- | --- | ---
| theme1   | Problem1   | Problem2    | ...     | Uncertain |
| theme1   | Problem1   | Problem2    | ...     | Uncertain |
| ...      | Problem1   | Problem2    | ...     | Uncertain |
| theme16  | Problem1   | Problem2    | ...     | Uncertain |

In fact, I repeat a lot of visually checking, definition locating, variables searching, and finally modifying for each theme, some of the work for many themes are exactly the same.

But until I nearly finish, I discover that I can actually make a list or set of all possible problems, `[problem1, problem2, problem3,..., problem50]`, then check these problems one by one through 16 themes, something like this:


|    | theme1  |  theme2 |  theme3 | ...      | theme16 
--- | --- | --- | --- | --- | ---
| problem1   |         | Modified|         | Modified |          |
| problem2   | Modified|         |         |          |          |
| problem3   | Modified|         | Modified|          |          |
| ...        |         |         |         | Modified |          |
| problem50  |         | Modified| Modified|          | Modified |

In this way, I check the same problem in every theme and modify the wrong one. So the big problem is to first find all possible problem, as your have said, the similar general problems for every theme. This seems better, and indeed sometimes, I have to go back and check the previous modified themes committed again for newly discovered problems. But, I did not write down every problems in the process, and waste some time.

I think this is a lesson for me, for problems that MAY appear repeatedly in the future, write down in a format, **how it happens, where it comes from, how it can be solved**. Why a problem will appear again? That's another question, actually I have not think it clearly enough at present.

xiaohanyu commented:
Wonderful job! Just got time to give a shot to this amazing PR.

No time to dive into the details, only some minor thoughts:
- Seems you use 4-spaces indentation, however, official semantic-ui uses 2-spaces indentation. As you said, we need some rules and lint tools to do code check here. The is a later issue.
- I've seen that you add some extra `warning`, `error` class to some element, which, is not exists in official semantic-ui. 

I will merge this PR right now, and maybe later we can start a third round with this solid foundational work, but this still need some extra discussion. 

Anyway, thanks for your job!

xiaohanyu commented:
And please help update this repo: https://github.com/semantic-ui-forest/themes.

Sorry I still didn't get time to write a automatic batch build script.

tarvos21 commented:
OK, I will update the built theme files.
